File: networktables/java/src/main/java/edu/wpi/first/wpilibj/networktables2/WriteManager.java

package edu.wpi.first.wpilibj.networktables2;

import edu.wpi.first.wpilibj.networktables2.thread.*;
import edu.wpi.first.wpilibj.networktables2.util.*;

/**
 * A write manager is a {@link IncomingEntryReceiver} that buffers transactions and then and then dispatches them to a flushable transaction receiver that is periodically offered all queued transaction and then flushed
 * 
 * @author Mitchell
 *
 */
public class WriteManager implements OutgoingEntryReceiver, PeriodicRunnable{
	private final int SLEEP_TIME = 100;
        
	private final int queueSize = 500;
	
	private Object transactionsLock = new Object();
	private NTThread thread;
	private NTThreadManager threadManager;
	private final AbstractNetworkTableEntryStore entryStore;
	
	private volatile HalfQueue incomingAssignmentQueue;
	private volatile HalfQueue incomingUpdateQueue;
	private volatile HalfQueue outgoingAssignmentQueue;
	private volatile HalfQueue outgoingUpdateQueue;
	
	private FlushableOutgoingEntryReceiver receiver;
	private long lastWrite;

	private final long keepAliveDelay;

	/**
	 * Create a new Write manager
	 * @param receiver
	 * @param threadManager
	 * @param transactionPool
	 * @param entryStore
	 */
	public WriteManager(final FlushableOutgoingEntryReceiver receiver, final NTThreadManager threadManager, final AbstractNetworkTableEntryStore entryStore, long keepAliveDelay) {
		this.receiver = receiver;
		this.threadManager = threadManager;
		this.entryStore = entryStore;
		
		incomingAssignmentQueue = new HalfQueue(queueSize);
		incomingUpdateQueue = new HalfQueue(queueSize);
		outgoingAssignmentQueue = new HalfQueue(queueSize);
		outgoingUpdateQueue = new HalfQueue(queueSize);
		
		this.keepAliveDelay = keepAliveDelay;
	}

	/**
	 * start the write thread
	 */
	public void start(){
		if(thread!=null)
			stop();
		lastWrite = System.currentTimeMillis();
		thread = threadManager.newBlockingPeriodicThread(this, "Write Manager Thread");
	}
	/**
	 * stop the write thread
	 */
	public void stop(){
		if(thread!=null)
			thread.stop();
	}


	public void offerOutgoingAssignment(NetworkTableEntry entry) {
		synchronized(transactionsLock){
			incomingAssignmentQueue.queue(entry);
			if(incomingAssignmentQueue.isFull()){
				try {
					run();
				} catch (InterruptedException e) {}
				System.err.println("assignment queue overflowed. decrease the rate at which you create new entries or increase the write buffer size");
			}
		}
	}


	public void offerOutgoingUpdate(NetworkTableEntry entry) {
		synchronized(transactionsLock){
			incomingUpdateQueue.queue(entry);
			if(incomingUpdateQueue.isFull()){
				try {
					run();
				} catch (InterruptedException e) {}
				System.err.println("update queue overflowed. decrease the rate at which you update entries or increase the write buffer size");
			}
		}
	}

	
	/**
	 * the periodic method that sends all buffered transactions
	 */
	public void run() throws InterruptedException {
		synchronized(transactionsLock){
			//swap the assignment and update queue
			HalfQueue tmp = incomingAssignmentQueue;
			incomingAssignmentQueue = outgoingAssignmentQueue;
			outgoingAssignmentQueue = tmp;
			
			tmp = incomingUpdateQueue;
			incomingUpdateQueue = outgoingUpdateQueue;
			outgoingUpdateQueue = tmp;
		}
		
		boolean wrote = false;
		NetworkTableEntry entry;
		int i;
		int size = outgoingAssignmentQueue.size();
		Object[] array = outgoingAssignmentQueue.array;
		for(i = 0; i<size; ++i){
			entry = (NetworkTableEntry)array[i];
			synchronized(entryStore){
				entry.makeClean();
			}
			wrote = true;
			receiver.offerOutgoingAssignment(entry);
		}
		outgoingAssignmentQueue.clear();
		
		
		size = outgoingUpdateQueue.size();
		array = outgoingUpdateQueue.array;
		for(i = 0; i<size; ++i){
			entry = (NetworkTableEntry)array[i];
			synchronized(entryStore){
				entry.makeClean();
			}
			wrote = true;
			receiver.offerOutgoingUpdate(entry);
		}
		outgoingUpdateQueue.clear();
		
		
		if(wrote){
			receiver.flush();
			lastWrite = System.currentTimeMillis();
                }
		else if(System.currentTimeMillis()-lastWrite>keepAliveDelay)
			receiver.ensureAlive();
		
		Thread.sleep(SLEEP_TIME);
	}

}

File: networktables/cpp/lib/share/networktables2/WriteManager.cpp
/*
 * WriteManager.cpp
 *
 *  Created on: Sep 25, 2012
 *      Author: Mitchell Wills
 */

#include "networktables2/WriteManager.h"
#include "networktables2/util/System.h"
#include "networktables2/AbstractNetworkTableEntryStore.h"
#include <iostream>



WriteManager::WriteManager(FlushableOutgoingEntryReceiver& _receiver, NTThreadManager& _threadManager, AbstractNetworkTableEntryStore& _entryStore, unsigned long _keepAliveDelay)
	: receiver(_receiver), threadManager(_threadManager), entryStore(_entryStore), keepAliveDelay(_keepAliveDelay){
	
	thread = NULL;
	lastWrite = 0;
	
	incomingAssignmentQueue = new std::queue<NetworkTableEntry*>();
	incomingUpdateQueue = new std::queue<NetworkTableEntry*>();
	outgoingAssignmentQueue = new std::queue<NetworkTableEntry*>();
	outgoingUpdateQueue = new std::queue<NetworkTableEntry*>();
}

WriteManager::~WriteManager(){
	stop();

	//Note: this must occur after stop() to avoid deadlock
	transactionsLock.take();
	
	delete incomingAssignmentQueue;
	delete incomingUpdateQueue;
	delete outgoingAssignmentQueue;
	delete outgoingUpdateQueue;
}

void WriteManager::start(){
	if(thread!=NULL)
		stop();
	lastWrite = currentTimeMillis();
	thread = threadManager.newBlockingPeriodicThread(this, "Write Manager Thread");
}

void WriteManager::stop(){
  if(thread!=NULL){
    thread->stop();
    delete thread;
    thread = NULL;
  }
}


void WriteManager::offerOutgoingAssignment(NetworkTableEntry* entry) {
	{ 
		bool test_queue_overflow=false;
		{
			NTSynchronized sync(transactionsLock);
			((std::queue<NetworkTableEntry*>*)incomingAssignmentQueue)->push(entry);

			test_queue_overflow=(((std::queue<NetworkTableEntry*>*)incomingAssignmentQueue)->size()>=queueSize);
		}
		if (test_queue_overflow)
		{
			run();
			writeWarning("assignment queue overflowed. decrease the rate at which you create new entries or increase the write buffer size");
		}
	}
}


void WriteManager::offerOutgoingUpdate(NetworkTableEntry* entry) {
	{ 
		bool test_queue_overflow=false;
		{
			NTSynchronized sync(transactionsLock);
			((std::queue<NetworkTableEntry*>*)incomingUpdateQueue)->push(entry);
			test_queue_overflow=(((std::queue<NetworkTableEntry*>*)incomingUpdateQueue)->size()>=queueSize);

		}
		if (test_queue_overflow)
		{
			run();
			writeWarning("update queue overflowed. decrease the rate at which you update entries or increase the write buffer size");
		}
	}
}


void WriteManager::run() {
	{
		NTSynchronized sync(transactionsLock);
		//swap the assignment and update queue
		volatile std::queue<NetworkTableEntry*>* tmp = incomingAssignmentQueue;
		incomingAssignmentQueue = outgoingAssignmentQueue;
		outgoingAssignmentQueue = tmp;
		
		tmp = incomingUpdateQueue;
		incomingUpdateQueue = outgoingUpdateQueue;
		outgoingUpdateQueue = tmp;
	}
	
	bool wrote = false;
	NetworkTableEntry* entry;
	
	while(!((std::queue<NetworkTableEntry*>*)outgoingAssignmentQueue)->empty()){
		entry = ((std::queue<NetworkTableEntry*>*)outgoingAssignmentQueue)->front();
		((std::queue<NetworkTableEntry*>*)outgoingAssignmentQueue)->pop();
		{
			NetworkTableEntry * entryCopy;

			{
				NTSynchronized sync(entryStore.LOCK);
				entry->MakeClean();
				wrote = true;
				entryCopy = new NetworkTableEntry(*entry);
			}

			receiver.offerOutgoingAssignment(entryCopy);
			delete entryCopy;
		}
	}
	
	while(!((std::queue<NetworkTableEntry*>*)outgoingUpdateQueue)->empty()){
		entry = ((std::queue<NetworkTableEntry*>*)outgoingUpdateQueue)->front();
		((std::queue<NetworkTableEntry*>*)outgoingUpdateQueue)->pop();
		{ 
			NetworkTableEntry * entryCopy;

			{
				NTSynchronized sync(entryStore.LOCK);
				entry->MakeClean();
				wrote = true;
				entryCopy = new NetworkTableEntry(*entry);
			}

			receiver.offerOutgoingUpdate(entryCopy);
			delete entryCopy;
		}
	}
	
	
	
	if(wrote){
		receiver.flush();
		lastWrite = currentTimeMillis();
	}
	else if(currentTimeMillis()-lastWrite>keepAliveDelay)
		receiver.ensureAlive();
	
	sleep_ms(20);
}

